### DISCUSSION TITLE
Image data selection for optimal performance in YOLO

### DISCUSSION THREAD

**Discussion by varungupta31** [AlexeyAB/darknet#7754] [Q&A]
I'm training a custom YOLOv4 and YOLOv4 tiny model. The object that I'm trying to track is human nose and have images of it at 2 distances -

- Further away from the camera (around 2600 images)
- Closer to the camera ( around 1600 images)

Though, the model will be used for detection at closer range,
I had a couple of questions regarding data selection for YOLO, for optimal results:

1. **What would be the best choice in the above case? (also why, so I can generalize the knowledge in the future)**
2. **If I train using all the images ( around 4200 images) for a single class detector (though at two different distances), will it lead to over-fitting?**
3. Besides these, what have been recommended practices for training image selection for YOLO (beyond the fact of 2000 images per class) 
4. And do these apply directly to YOLOv4 tiny, or there some things I should keep mind with the smaller model.
(Does is it work better if fed more images compared to YOLOv4 regular?, etc.)

Lastly, the object in-focus in this thread is a small object, but are these concepts similar for medium and large size objects too?
(Or there are some other points to keep in mind with those type of detectors)

**P.s. - My priority is to understand the first two questions, but please pitch-in if you have anything to contribute to the discussion.**

Thanks in advance!

**Comment by stephanecharette** [comment]
If  you only want to track the noses close up, and *never* want to track the nose from far away, then why train with the far away images?  However -- what about halfway in between?  What if they're walking up to the camera?  Or they're leaning back?  From that point of view it makes sense to include all the images.

To consider:  are all the noses your nose?  Or very diverse noses from hundreds of people?  What if they have glasses on?  What if they have "Halloween" style makeup on with lots of different colours?  What if they have a "clown" nose?  What about all the various skin tones?  What about people who turn their head?

Have you considered negative samples and other scenarios?  For example, you should also have the following pictures:

1) people who have turned around to look backwards, so now you only see the back of their head
2) people who are looking up at the ceiling, so now you see their neck, chin, or maybe a view up their nose
3) people who are looking at the floor or their lap, so now you see the top of their head
4) people who have long hair covering their face, so you cannot see their nose
5) people who have stepped away from their desk so the image is a view of the room without a face
6) multiple people in the view frame, so now you have 2, 3, or more noses in the image
7) lights off in the room

All of this applies to both yolo and yolo-tiny.  And there is absolutely zero reason to use anything other than yolov4-tiny for the scenario you describe.

> the object in-focus in this thread is a small object

How so?  Are we not talking about a webcam type of image where the face take up a large part of the image?  If so, then this does not qualify as "small object" unless you tell me you annotated the two nostrils.

If you have 1 class ("nose"), and you are using yolov4-tiny, and you've annotated your images correctly, I doubt you'll even need the full 6000 minimum for max batches.  This should be a relatively simple neural network to train, and you should see some decent results relatively quickly into the training.

**Reply by varungupta31** [reply]
Thank you for the elaborate response, I highly appreciate your inputs. 

- I do want the flexibility mentioned, so will be including all the images.

> webcam type of image where the face take up a large part of the image

- Got it, so it's not a small object.

- I do have different subjects - covering multiple orientations. 

>  If I train using all the images ( around 4200 images) for a single class detector (though at two different distances), will it lead to over-fitting?


My concern regarding over-fitting arose because earlier I was training a model for a driver using their phone and used the [state farm distracted driver dataset](https://www.kaggle.com/c/state-farm-distracted-driver-detection/data). It contained images of person using their phones in both hands, at various eye levels and at different orientation (around 2500 images), but it seemed to work very well when tested on the train data (even images where the phone was barely visible, the model could detect with high prob! but when given new images, it worked terribly poor) 

I understand without you taking a look at the data,  you can't comment on why this happened - but isn't this a case of over-fitting? and is there any other possibility that could have caused this, except a flawed train data (though I clearly remember running it on similar (to train data) but new images and still resulting in no/poor detection) ?

Lastly,

> I doubt you'll even need the full 6000 minimum for max batches

Is there any 'earlystopping' support in YOLO? Or a way to track the training progress in real-time? (I use Google Colab where the training chart can be viewed after the training is complete)

Thanks again :)

**Reply by stephanecharette** [reply]
I know nothing about your data set or why some other project you made in the past may have failed to work well.  You'll have to experiment and see.

> Or a way to track the training progress in real-time?

The chart.png file is written as training happens, it isn't only when the training ends.  You can view it at any point in time and stop training as you see fit.  Make sure you include `-map` and `-dont_show` in your parameters, as described here: https://www.ccoderun.ca/programming/darknet_faq/#training_command
